### DISCUSSION TITLE
MooseMesh: property "add_subdomain_names"?

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#27415] [Q&A General] [answered]
Dear Community,

Following the procedure recommended [here](https://mooseframework.inl.gov/framework/contributing.html), I would like to propose the property "add_subdomain_names" for the object "MooseMesh" for discussion.

**Motivation and current situation**
Using the property `add_subdomain_ids` of the [MooseMesh base class](https://mooseframework.inl.gov/source/mesh/MooseMesh.html) a user can add (initially empty) subdomains to an existing mesh for later usage. Up to my knowledge today only subdomain IDs can be added and no name is provided for these subdomains. To ease reading and debugging of the Moose input file it would be great to also provide the possibility to add subdomain names. For this purpose I like to propose an additional property for the object "MooseMesh": add_subdomain_names

**Proposed property "add_subdomain_names" for the object MooseMesh**
Following the design pattern of existing objects and their properties, e.g. with "symbol_names" and "symbol_values" in [ParsedFunction](https://mooseframework.inl.gov/source/functions/MooseParsedFunction.html) I like to add the property `add_subdomain_names` as a companion property to the existing property `add_subdomain_ids`. While `add_subdomain_ids` is taking a vector of subdomain ids, the additional (optional) `add_subdomain_names` provides the corresponding vector of SubdomainNames. The property add_subdomain_names would only be valid, if add_subdomain_ids is defined.

A input file using this property might look like:
```
[Mesh]
  ...
  add_subdomain_ids  =  '5      6      7'
  add_subdomain_names = 'BlockX BlockY BlockZ'
  ...
[]
```

**Impact to existing Moose apps**
As this proposal only adds functionality. Existing Moose apps should be able to cope with the proposed changes without any problems.

Please find a crude proof of concept [here](https://github.com/grunerjmeier/moose/blob/279ee92f19f34c3c9ebed262898e748d25af1414/framework/src/mesh/MooseMesh.C#L169) and [there](https://github.com/grunerjmeier/moose/blob/279ee92f19f34c3c9ebed262898e748d25af1414/framework/src/mesh/MooseMesh.C#L409).

If this proposal meets with interest, I need support in preparing the code before I can create a PR. The proof-of-concept works for me, but does not include support for "parallel mesh", for example.

I would be delighted if you could consider this proposal.

**Comment by GiudGiud** [comment]
Hello

This is a good idea. 

I think we will want to match the number and names and ids 
OR have names specified and no ids
OR have ids specified and no ids

I'll tag the original contributor of the id feature on the PR for review

Guillaume

**Reply by jmeier** [reply]
Dear Guillaume,

Thanks for your quick and positive reply. 

> I think we will want to match the number and names and ids

This is how I implemented it in my proof of concept linked above.

> OR have names specified and no ids

Then ids need to be generated on-the-fly. I would need help to implement that.

> OR have ids specified and no ids _(I think you mean "names")_

To just provide ids (and NO names) is possible in my proof of concept linked above and I think it is crucial to maintain this behavior. Otherwise we would break some of the existing models.

Please let me know when it is clear how to proceed. There are still a few things in my proof of concept that I am unsure about. Should these questions be asked here - or as part of a PR?

**Reply by GiudGiud** [reply]
I think a PR is the way forward if you have something working already

**Reply by jmeier** [reply]
Please find the corresponding PR here: [#27420](https://github.com/idaholab/moose/pull/27420)

**Comment by jmeier** [comment, marked as answer]
The PR is now merged. Thanks to everyone for your support!
